java/ql/src/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -82,44 +82,107 @@
 import java
 private import Completion
 
-/** A node in the expression-level control-flow graph. */
-class ControlFlowNode extends Top, @exprparent {
-  /** Gets the statement containing this node, if any. */
-  Stmt getEnclosingStmt() {
-    result = this or
-    result = this.(Expr).getEnclosingStmt()
-  }
+module ControlFlow {
+  private newtype TNode =
+    TExprNode(Expr e) { hasControlFlow(e) } or
+    TStmtNode(Stmt s) or
+    TExitNode(Callable c) { exists(c.getBody()) }
```

Comment:
C# also has an entry node, for symmetry.

Reply:
C# also uses a separate IPA type for basic blocks.

Reply:
I was aiming for minimal changes, so I'm not going to change the CFG with e.g. a new entry node here.  Using a separate type for basic blocks could of course be done, but I don't think it's that important, so I think I'll leave it as is for now.